pyomo/opt/solver/shellcmd.py

```diff
@@ -299,14 +299,20 @@ def _execute_command(self,command):
                 _input = command.script
             else:
                 _input = None
-            [rc, log] = run(
+            result = subprocess.run(
                 command.cmd,
                 stdin = _input,
-                timelimit = self._timelimit if self._timelimit is None else self._timelimit + max(1, 0.01*self._timelimit),
+                timeout = self._timelimit if self._timelimit is None else self._timelimit + max(1, 0.01*self._timelimit),
                 env   = command.env,
-                tee   = self._tee,
-                define_signal_handlers = self._define_signal_handlers
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+                universal_newlines=True
              )
+            if self._tee:
+                for line in result.stdout:
+                    sys.stdout.write(line)
```

Comment:
Can we have a bigger conversation on this? It looks like the [section that does this](https://github.com/PyUtilib/pyutilib/blob/master/pyutilib/subprocess/processmngr.py#L587) in pyutilib is based on threading, and it's quite long.

Reply:
Reading more on the vast interwebs, it looks like threading is the only way to do this to produce the output as desired. How would you like to proceed?

Reply:
I'm think I'm still not convinced that the solution I developed here (https://github.com/michaelbynum/pyomo/blob/appsi/pyomo/contrib/appsi/utils/tee_thread.py) and here (https://github.com/michaelbynum/pyomo/blob/47beab745db7f4601aaac3e247b15f7ef434a82b/pyomo/contrib/appsi/solvers/ipopt.py#L277) is not sufficient. I think we discussed a problem where the last few lines of the log file might not be printed to stdout, but I think that can be easily fixed by giving `TeeThread` a file object instead of a filename. Then put subprocess in a try-finally block where the finally clause makes sure everything was written. Maybe there are other issues that I am forgetting now.

Reply:
There are a number of issues here.  I will try to (somewhat) succinctly summarize what I have been thinking.

First, the bulk of this PR is good (we want to use Python's `subprocess` directly whenever possible).  However, there are a couple cases (notably the solver subprocesses) where we want to be able to send stdout/stderr to more than one location.  This includes 1) to the screen (in real time) for the user's benefit, 2) to a log file for archival purposes, and 3) [for some solvers] to a string buffer for subsequent parsing to get relevant results information back (like iteration count from Ipopt).  After some Googling, it appears that there is no way to do this without using either `threading` or `asyncio`.  For these situations I feel that we should have a separate utility (*not* called `subprocess`) that wraps `subprocess` handles the "tee" activity.

@michaelbynum has suggested a (beautifully simple) approach whereby we don't actually "tee", but instead send the stream to a file and have a separate thread that basically does `tail -f`.  This has the benefit of being the simplest implementation that I have run across.  However, it has some limitations.  The biggest is that it is tied to the filesystem.  In other projects (notably Dakota where inter-process communication can occur through the filesystem) we have seen numerous challenges when writing and then immediately reading from files.  This is particularly true on HPC systems where the filesystem is NFS or Lustre.  I forget the exact root cause, but I recall it being something like all the file data was not available, even after the process terminates.  It had to do with the (significant) buffering that occurs (both in the file descriptors and the OS).  It is aggravated by filesystems that are optimized for writing large amounts of data, but not for write/read operations (like Lustre).  There is also the principle that we really shouldn't be creating files if we don't actually need a file.

The solutions that are more akin to `tee` have the benefit of being entirely in-memory: so if you are in a situation where the user did not ask for the log file, but you need to parse the solver log, then you only create the string buffer and dump things to the console: no files are created, so there is nothing to clean up.

A separate (but related) question is how we should handle merging stdout/stderr.  If we let the underlying OS handle the merge (i.e., by using `stderr=subprocess.STDOUT`), then you get the (unfortunately common) situation where a stderr message gets stuck in the middle of a line of text sent to stdout.  This makes parsing the resulting combined stream very challenging (both for automated parsers trying to extract information, and for users).  A lot of the work in the PyUtilib implementation was geared around making the merge process more reliable and easier to parse (in particular, it tried to ensure that entire lines were kept together).

Finally, there are libraries out there other than PyUtilib that (mostly?) solve this problem.  As an alternative to rolling our own, we could add another dependency to Pyomo.

Reply:
The idea of adding another dependency that we do not maintain is appealing to me. We are ripping out PyUtilib and Enum in 6.0.0, which means Pyomo still only requires:
```
      install_requires=[
          'ply',
          'six>=1.4',
      ],
```
Plus whatever library you think might be a good alternative, @jsiirola .

Reply:
I think the challenge of not all of the data being in the file can be overcome pretty easily, but I'm not entirely sure.

Are we sure that subprocess does not directly handle the correct merging of stdout/stderr?

Reply:
Forget my last comment about merging stdout/stderr. I understand.

Reply:
The correct merging of stdout/stderr is certainly more difficult, but I'm not convinced that it is terribly important (yet).